compiler/src/dotty/tools/dotc/inlines/InlineReducer.scala

```diff
@@ -158,35 +158,46 @@ class InlineReducer(inliner: Inliner)(using Context):
+    *
+    *  This variant of beta-reduction preserves the integrity of `Inlined` tree nodes.
     */
   def betaReduce(tree: Tree)(using Context): Tree = tree match {
-    case Apply(Select(cl @ closureDef(ddef), nme.apply), args) if defn.isFunctionType(cl.tpe) =>
-      // closureDef also returns a result for closures wrapped in Inlined nodes.
-      // These need to be preserved.
-      def recur(cl: Tree): Tree = cl match
-        case Inlined(call, bindings, expr) =>
-          cpy.Inlined(cl)(call, bindings, recur(expr))
-        case _ => ddef.tpe.widen match
-          case mt: MethodType if ddef.paramss.head.length == args.length =>
-            val bindingsBuf = new DefBuffer
-            val argSyms = mt.paramNames.lazyZip(mt.paramInfos).lazyZip(args).map { (name, paramtp, arg) =>
-              arg.tpe.dealias match {
-                case ref @ TermRef(NoPrefix, _) => ref.symbol
-                case _ =>
-                  paramBindingDef(name, paramtp, arg, bindingsBuf)(
-                    using ctx.withSource(cl.source)
-                  ).symbol
+    case Apply(Select(cl, nme.apply), args) if defn.isFunctionType(cl.tpe) =>
+      val bindingsBuf = new DefBuffer
+      def recur(cl: Tree): Option[Tree] = cl match
+        case Block((ddef : DefDef) :: Nil, closure: Closure) if ddef.symbol == closure.meth.symbol =>
+          ddef.tpe.widen match
+            case mt: MethodType if ddef.paramss.head.length == args.length =>
+              val argSyms = mt.paramNames.lazyZip(mt.paramInfos).lazyZip(args).map { (name, paramtp, arg) =>
+                arg.tpe.dealias match {
+                  case ref @ TermRef(NoPrefix, _) => ref.symbol
+                  case _ =>
+                    paramBindingDef(name, paramtp, arg, bindingsBuf)(
+                      using ctx.withSource(cl.source)
+                    ).symbol
+                }
               }
-            }
-            val expander = new TreeTypeMap(
-              oldOwners = ddef.symbol :: Nil,
-              newOwners = ctx.owner :: Nil,
-              substFrom = ddef.paramss.head.map(_.symbol),
-              substTo = argSyms)
-            Block(bindingsBuf.toList, expander.transform(ddef.rhs)).withSpan(tree.span)
-          case _ => tree
-      recur(cl)
-    case _ => tree
+              val expander = new TreeTypeMap(
+                oldOwners = ddef.symbol :: Nil,
+                newOwners = ctx.owner :: Nil,
+                substFrom = ddef.paramss.head.map(_.symbol),
+                substTo = argSyms)
+              Some(expander.transform(ddef.rhs))
+            case _ => None
+        case Block(stats, expr) if stats.forall(isPureBinding) =>
+          recur(expr).map(cpy.Block(cl)(stats, _))
+        case Inlined(call, bindings, expr) if bindings.forall(isPureBinding) =>
+          recur(expr).map(cpy.Inlined(cl)(call, bindings, _))
+        case Typed(expr, tpt) =>
+          recur(expr)
```

Comment:
Any reason we don't preserve the Typed node here?
```suggestion
          recur(expr).map(cpy.Typed(cl)(_, tpt))
```

Reply:
It is intentional. This ascription is on the lambda and hence has the lambda type, but the result of this transformation has the type of the result of this lambda. Dropping it is the simplest option. We could attempt to add the ascription again on the result. This would be more complicated and would also prevent optimizations on constants. We also do the same in https://github.com/lampepfl/dotty/blob/main/compiler/src/dotty/tools/dotc/transform/BetaReduce.scala#L57-L58.